src/data_source/storage/sql/queries/explorer.rs

```diff
@@ -509,61 +513,76 @@ where
         &mut self,
-        search_query: String,
+        search_query: TaggedBase64,
     ) -> Result<SearchResult<Types>, GetSearchResultsError> {
-        let block_query = format!(
-            "SELECT {BLOCK_COLUMNS}
-                FROM header AS h
-                JOIN payload AS p ON h.height = p.height
-                WHERE h.hash LIKE $1 || '%'
-                ORDER BY h.height DESC
-                LIMIT 5"
-        );
-        let block_query_rows = query(block_query.as_str())
-            .bind(&search_query)
-            .fetch(self.as_mut());
-        let block_query_result: Vec<BlockSummary<Types>> = block_query_rows
-            .map(|row| BlockSummary::from_row(&row?))
-            .try_collect()
-            .await?;
+        let search_tag = search_query.tag();
+        let header_tag = Commitment::<Header<Types>>::tag();
+        let tx_tag = Commitment::<HotshotTransaction<Types>>::tag();
 
-        let transactions_query = format!(
-            "SELECT {BLOCK_COLUMNS}
-                FROM header AS h
-                JOIN payload AS p ON h.height = p.height
-                JOIN transaction AS t ON h.height = t.block_height
-                WHERE t.hash LIKE $1 || '%'
-                ORDER BY h.height DESC
-                LIMIT 5"
-        );
-        let transactions_query_rows = query(transactions_query.as_str())
-            .bind(&search_query)
-            .fetch(self.as_mut());
-        let transactions_query_result: Vec<TransactionSummary<Types>> = transactions_query_rows
-            .map(|row| -> Result<Vec<TransactionSummary<Types>>, QueryError>{
-                let block = BlockQueryData::<Types>::from_row(&row?)?;
-                let transactions = block
-                    .enumerate()
-                    .enumerate()
-                    .filter(|(_, (_, txn))| txn.commit().to_string().starts_with(&search_query))
-                    .map(|(offset, (_, txn))| {
-                        Ok(TransactionSummary::try_from((
-                            &block, offset, txn,
-                        ))?)
-                    })
-                    .try_collect::<TransactionSummary<Types>, Vec<TransactionSummary<Types>>, QueryError>()?;
+        if search_tag != header_tag && search_tag != tx_tag {
+            return Err(GetSearchResultsError::InvalidQuery(errors::BadQuery {}));
+        }
 
-                Ok(transactions)
+        let search_query_string = search_query.to_string();
+        if search_tag == header_tag {
+            let block_query = format!(
+                "SELECT {BLOCK_COLUMNS}
+                    FROM header AS h
+                    JOIN payload AS p ON h.height = p.height
+                    WHERE h.hash = $1
+                    ORDER BY h.height DESC
+                    LIMIT 5"
+            );
+            let block_query_rows = query(block_query.as_str())
+                .bind(&search_query_string)
+                .fetch(self.as_mut());
```

Comment:
We can do `LIMIT 1` and `fetch_one` here, querying for blocks by equality on hash is only ever going to return one block

Reply:
I am considering making the blocks field an Option<BlockSummary> since there can only be one block now. However, I don't want to break the frontend, and I also think that SearchResult should be an enum

Reply:
Making this return an `enum` would indeed be a backwards incompatible change.  I can possibly support it with a backwards compatible decoder that can support either representation, but I think that would need to be in place in the Block Explorer client-side code before that sort of change is introduced on the server side.